src/Lean/Elab/Term.lean

```diff
@@ -933,51 +932,80 @@ private def postponeElabTerm (stx : Syntax) (expectedType? : Option Expr) : Term
   registerSyntheticMVar stx mvar.mvarId! (SyntheticMVarKind.postponed (← saveContext))
   pure mvar
 
+def getSyntheticMVarDecl? (mvarId : MVarId) : TermElabM (Option SyntheticMVarDecl) :=
+  return (← get).syntheticMVars.find? fun d => d.mvarId == mvarId
+
+def mkTermInfo (stx : Syntax) (e : Expr) (expectedType? : Option Expr := none) : TermElabM (Sum Info MVarId) := do
+  let isHole? : TermElabM (Option MVarId) := do
+    match e with
+    | Expr.mvar mvarId _ =>
+      match (← getSyntheticMVarDecl? mvarId) with
+      | some { kind := SyntheticMVarKind.tactic .., .. }    => return mvarId
+      | some { kind := SyntheticMVarKind.postponed .., .. } => return mvarId
+      | _                                                   => return none
+    | _ => pure none
+  match (← isHole?) with
+  | none        => return Sum.inl <| Info.ofTermInfo { elaborator := (← MonadRef.getElaborator), lctx := (← getLCtx), expr := e, stx, expectedType? }
+  | some mvarId => return Sum.inr mvarId
+
+def addTermInfo (stx : Syntax) (e : Expr) (expectedType? : Option Expr := none) : TermElabM Unit := do
+  withInfoContext' (pure ()) (fun _ => mkTermInfo stx e expectedType?) |> discard
+
 /-
   Helper function for `elabTerm` is tries the registered elaboration functions for `stxNode` kind until it finds one that supports the syntax or
   an error is found. -/
 private def elabUsingElabFnsAux (s : SavedState) (stx : Syntax) (expectedType? : Option Expr) (catchExPostpone : Bool)
-    : List TermElab → TermElabM Expr
+    : List (KeyedDeclsAttribute.AttributeEntry TermElab) → TermElabM Expr
```

Comment:
How confident are you about the changes here?
This code is very delicate. I will have to read it carefully.
A few comments motivating the changes will help me to review the code.

Reply:
For example, the new code has two nested `try-catch`. why?

Reply:
I tried to explain the double try in the new comments. I don't think there are any changes other than that and moving the withInfoContext'.